arduino.go

```diff
@@ -0,0 +1,156 @@
+// Package arduino helps to talk to
+// an Arduino connected via USB (to a Raspberry Pi).
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"os"
+
+	"context"
+	"github.com/pkg/errors"
+	"github.com/spf13/afero"
+	"strings"
+	"sync"
+)
+
+var (
+	// AppFs is an abstraction of the file system
+	// to allow mocking in tests.
+	AppFs = afero.NewOsFs()
+)
+
+const (
+	// ReceiveCmd is the command that needs to b e sent to the Arduino to start
+	// receiving on Pin 0.
+	ReceiveCmd = "RF receive 0"
+	// ReceivePrefix is the prefix of raw signals read from the device file of the Arduino.
+	ReceivePrefix = "RF receive "
+
+	// ResetCmd is resetting the homeduino device to set it in an expected state
+	ResetCmd = "RESET"
+
+	// ResetCmdResponse is the response to ResetCmd homeduiono
+	ResetCmdResponse = "ready"
+)
+
+// ProcessorFunc are implementations for consuming line by line of the device
+type ProcessorFunc func(s string) bool
+
+// Device represents the device file of an Arduino
+// connected to the USB port.
+type Device struct {
+	afero.File
+	sync.Mutex
+	open bool
+}
+
+// OpenDevice opens the named device file for reading.
+func OpenDevice(name string) (*Device, error) {
+	file, err := AppFs.OpenFile(name, os.O_RDWR, 0644)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to open '%v'", name)
+	}
+	log.Printf("Device '%v' opened", file.Name())
+
+	d := &Device{
+		File:  file,
+		open:  true,
+		Mutex: sync.Mutex{},
+	}
+
+	return d, nil
+}
+
+// Reset sends ResetCmd to the device to set it in an expected state
+func (d *Device) Reset() error {
+	err := d.Write(ResetCmd)
+	if err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	err = d.Process(ctx, func(s string) bool {
+		if !strings.Contains(s, ResetCmdResponse) {
+			log.Printf("Not %v msg: %v\n", ResetCmdResponse, s)
+			return false
+		}
+		return true
+	})
+
+	if err != nil && err != context.Canceled {
+		return err
+	}
+
+	return nil
+}
+
+// Process reads the next line from a device file in a loop and
+// applies a ProcessorFunc to it. The Context is used to stop reading
+// Before ReadProcess can be used the device file needs to be opened via Open.
+func (d *Device) Process(ctx context.Context, handle ProcessorFunc) error {
+	d.Lock()
+	defer d.Unlock()
+	if !d.open {
+		return errors.New("File already closed")
+	}
+
+	scanner := bufio.NewScanner(d)
+	for scanner.Scan() {
+		line := scanner.Text()
+		log.Println("Line Scanned:", line)
+
+		stop := handle(line)
+		if stop {
+			return nil
+		}
+
+		select {
+		case <-ctx.Done():
```

Comment:
It was meant to be just there for being able to Cancel or End the Processing if needed. But you are right.. it is not really used right now. 
I just Context by now is kind of a deafult in go to be able to stop things and keep control.

I think it could be done better then it is right now TBH.